### ISSUE TITLE
bug： nni/algorithms/compression/v2/pytorch

### ISSUE THREAD

**Issue by felixfuu** [microsoft/nni#4221]
https://github.com/microsoft/nni/blob/master/nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py#L106

https://github.com/microsoft/nni/blob/87001025a4f4684e0458d990e517bc58b83ae10f/nni/compression/pytorch/utils/shape_dependency.py#L93

**Comment by J-shang** [comment]
@felixfuu Thanks~ This will be fixed in #4224

**Comment by felixfuu** [comment]
bug: https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L164 
when Prelu(num_parameters=1) @J-shang

**Comment by felixfuu** [comment]
> @felixfuu Thanks~ This will be fixed in #4224

I read your fix, there may be a new problem. There are two PrunerModuleWrapper in the code：
https://github.com/microsoft/nni/blob/master/nni/algorithms/compression/v2/pytorch/base/pruner.py#L18
https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/compressor.py#L278
, and an error will be reported in https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/utils/shape_dependency.py#L393， so maybe you should delete PrunerModuleWrapper(algorithms/compression/v2/pytorch/base/pruner.py) and keep PrunerModuleWrapper  in nni/compression/pytorch/compressor.py .

**Comment by felixfuu** [comment]
replace_batchnorm1d and replace_batchnorm2d have the same problem. When affine=False, there is no weight parameter, so norm.weight.data will be None, and https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L224, https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L267 will report an AttributeError.

**Comment by felixfuu** [comment]
@J-shang    Yes, I think the compression V2 is more reasonable, and I look forward to migrating all compressed code to V2.

some fix：
fix replace_prelu:
https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L112
`
weight_mask = weight_mask['weight']
if weight_mask.size(0) == 1:
    return prelu
`

fix  replace_batchnorm2d:
https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L266
`
# assign weights
    if norm.affine:
        new_norm.weight.data = torch.index_select(norm.weight.data, 0, remained_in)
        new_norm.bias.data = torch.index_select(norm.bias.data, 0, remained_in)
`

fix replace_batchnorm1d:
https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L223
`
    # assign weights
    if norm.affine:
        new_norm.weight.data = torch.index_select(norm.weight.data, 0, remained_in)
        new_norm.bias.data = torch.index_select(norm.bias.data, 0, remained_in)
`

**Comment by felixfuu** [comment]
> > > @felixfuu Thanks~ This will be fixed in #4224
> > 
> > 
> > I read your fix, there may be a new problem. There are two PrunerModuleWrapper in the code： https://github.com/microsoft/nni/blob/master/nni/algorithms/compression/v2/pytorch/base/pruner.py#L18 https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/compressor.py#L278 , and an error will be reported in [https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/utils/shape_dependency.py#L393，](https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/utils/shape_dependency.py#L393%EF%BC%8C) so maybe you should delete PrunerModuleWrapper(algorithms/compression/v2/pytorch/base/pruner.py) and keep PrunerModuleWrapper in nni/compression/pytorch/compressor.py .
> 
> @felixfuu oh, yes, seems you are trying to use compression v2, and I think these bugs have been fixed in #4125. you can try on this branch, and I will close #4224.
> 
> > replace_batchnorm1d and replace_batchnorm2d have the same problem. When affine=False, there is no weight parameter, so norm.weight.data will be None, and https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L224, https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/compress_modules.py#L267 will report an AttributeError.
> 
> and this seems a bug, you can make a pr to fix this issue if you have time. BTW, @zheng-ningxin do we support if `affine=False`?

#4125 commit： https://github.com/microsoft/nni/pull/4125/commits/d3df0b910e0f195a52d9905659f12a77b77d2c3d 
It seems that the PrunerModuleWrapper problem is still not fixed， this line
https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/utils/shape_dependency.py#L393 will always be False

**Comment by J-shang** [comment]
> #4125 commit： [d3df0b9](https://github.com/microsoft/nni/commit/d3df0b910e0f195a52d9905659f12a77b77d2c3d) It seems that the PrunerModuleWrapper problem is still not fixed， this line https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/utils/shape_dependency.py#L393 will always be False. Maybe you should delete PrunerModuleWrapper(algorithms/compression/v2/pytorch/base/pruner.py) and keep PrunerModuleWrapper in nni/compression/pytorch/compressor.py .

Hi @felixfuu,
seems you fetch an old commit, you can refer to this file `nni/compression/pytorch/utils/shape_dependency.py` in the following pr. I have tested it and it works.
https://github.com/microsoft/nni/pull/4125/files

**Comment by J-shang** [comment]
> https://github.com/microsoft/nni/blob/master/nni/algorithms/compression/v2/pytorch/pruning/tools/sparsity_allocator.py#L142
> 
> `python metric - metric.min()`
> 
> @J-shang This may be redundant, because the minimum index in metric may not be in the group_mask, then it is equivalent to one more 0, and the corresponding channel is pruned.

This is mainly to prevent the metric can be negative (even though we don't have this kind of metric right now). For example,
```
old_mask = [1, 1, 1, 0, 1]
new_metric = [-1, 2, 3, -2, -3]
sparsity = 0.4

# if we don't have metric - metric.min(), the new_mask may not inherit the old_mask
masked_metric = new_metric * old_mask = [-1, 2, 3, 0, -3]
new_mask = [0, 1, 1, 1, 0]

#  if we have metric - metric.min(), this problem can be solved
masked_metric = (new_metric - (-3)) * old_mask = [2, 5, 6, 0, 0]
new_mask = [1, 1, 1, 0, 0]
```

And you are right,  the minimum index in metric may not be in the group_mask. In the above example, if the sparsity is 0.2, there will be an error, so I think this place can be done as `metric - metric.min() + 1` to make sure the unmasked metric is bigger than 1, and the masked metric is equal to 0.

**Comment by felixfuu** [comment]
Can you update the QR code of WeChat? The current one is invalid. I want to join the WeChat group to facilitate communication.

**Comment by felixfuu** [comment]
> > @J-shang https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/infer_mask.py#L349
> > `
> > self.module.load_state_dict(self.state_dict) self.__apply_weight_mask() out = self.module(*self.dummy_input).clone().detach() if isinstance(out, torch.Tensor): constant = torch.zeros_like(out) constant_pos = out_mask == 0 constant[constant_pos] = out[constant_pos] `
> > I don't understand the purpose of this code. If you do self.__apply_weight_mask(), then out[constant_pos] will be 0. So constant[constant_pos] is always 0.
> 
> Masked weight may not case 0 output.
> 
> ```
> weight = [[2, 2], [2, 2]]
> weight_mask = [[1, 0], [1, 1]]
> input = [3, 3]
> output = [6, 12]
> ```
> 
> > Can you update the QR code of WeChat? The current one is invalid. I want to join the WeChat group to facilitate communication.
> 
> Could you give me a contact way, WeChat group is over 200 people and can not use QR code, I can invite you to join.
> 
> > also, here is a bug: https://github.com/microsoft/nni/blob/master/nni/compression/pytorch/speedup/infer_mask.py#L112, The name has been reset, which is not the original meaning of the code.
> 
> Thank you~ will fix it next release.

wechat: fcm_hust
